pkg/server/apiextensions.go

```diff
@@ -262,117 +277,238 @@ func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector
-	systemCRDKeyName := clusters.ToClusterAwareKey(SystemCRDLogicalCluster, name)
-	if systemCRDKeys.Has(systemCRDKeyName) {
-		crd, err = c.crdLister.Get(systemCRDKeyName)
-		if err != nil && !apierrors.IsNotFound(err) {
-			// something went wrong w/the lister - could only happen if meta.Accessor() fails on an item in the store.
-			return nil, err
+	return crd, nil
+}
+
+// shallowCopyCRD makes a shallow copy of in, with a deep copy of in.ObjectMeta.Annotations.
+func shallowCopyCRD(in *apiextensionsv1.CustomResourceDefinition) *apiextensionsv1.CustomResourceDefinition {
+	out := *in
+
+	out.Annotations = make(map[string]string)
+
+	for k, v := range in.Annotations {
+		out.Annotations[k] = v
+	}
+
+	return &out
+}
+
+// partialMetadataCRD modifies CRD and replaces all version schemas with minimal ones suitable for partial object
+// metadata.
+func partialMetadataCRD(crd *apiextensionsv1.CustomResourceDefinition) {
+	crd.Annotations[annotationKeyPartialMetadata] = ""
+
+	// set minimal schema that prunes everything but ObjectMeta
+	for _, v := range crd.Spec.Versions {
+		v.Schema = &apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type: "object",
+			},
 		}
+	}
+}
+
+func (c *apiBindingAwareCRDLister) getForFullData(clusterName logicalcluster.LogicalCluster, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	if clusterName == logicalcluster.Wildcard {
+		return c.getWildcard(name)
+	}
+
+	return c.get(clusterName, name)
+}
+
+func (c *apiBindingAwareCRDLister) getWildcard(name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	// TODO use an index by group+resource
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	crd, equivalentSchemas := findCRD(name, crds)
+	if !equivalentSchemas {
+		err = apierrors.NewInternalError(fmt.Errorf("error resolving resource: cannot watch across logical clusters for a resource type with several distinct schemas"))
+		return nil, err
+	}
+
+	if crd == nil {
+		return nil, errors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	}
+
+	return crd, nil
+}
+
+// getForIdentity handles finding the right CRD for an incoming wildcard request with identity, such as
+// /clusters/*/apis/$group/$version/$resource:$identity.
+func (c *apiBindingAwareCRDLister) getForIdentity(name, identity string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	group, resource := crdNameToGroupResource(name)
+
+	indexKey := apibinding.IdentityGroupResourceKeyFunc(identity, group, resource)
+
+	apiBindings, err := c.apiBindingIndexer.ByIndex(apibinding.IndexAPIBindingsByIdentityGroupResource, indexKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(apiBindings) == 0 {
+		return nil, errors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	}
 
-		if crd != nil {
-			return crd, nil
+	apiBinding := apiBindings[0].(*apisv1alpha1.APIBinding)
```

Comment:
this is the point where we could add gcd logic of some kind? Worth a TODO maybe.

Reply:
Added